### PR TITLE
Steve - L123.bcoc_awb_R_S_T_Y

### PR DESCRIPTION
**Pull request by ssmithClimate**
#129

### CODE REVIEW

File: R/constants.R [hunk 1]
```diff
@@ -95,6 +95,7 @@ CONV_HA_BM2 <- 1e-5
 CONV_THA_KGM2 <- 0.1   # tons C/ha -> kg C/m2
 CONV_GG_TG <- 0.001 # gigagrams to tegagrams
 CONV_TST_TG <- 0.000907 # thousand short tons to Tg
+CONV_KG_TO_TG <- 0.000000001
```
Comment (bpbond) [inline R/constants.R]: `1e-x` might clearer?

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
 #' the generated outputs: \code{L123.bcoc_tgmt_R_awb_2000}. The corresponding file in the
 #' original data system was \code{L123.bcoc_awb_R_S_T_Y.R} (emissions level1).
-#' @details Describe in detail what this chunk does.
+#' @details Use year 2000 BC and OC AWB emissions to calculate emission factors for each ag production technology.
 #' @importFrom assertthat assert_that
 #' @importFrom dplyr filter mutate select
 #' @importFrom tidyr gather spread
-#' @author YourInitials CurrentMonthName 2017
+#' @author SJS May 2017
 #' @export
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Remove `@export` line

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 3]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      select(iso, awb) %>%
+      mutate(Non.CO2 = "BC_AWB") -> BC_AWB_2000
+
+    RCP_OC_2000 %>%
+      select(iso, awb) %>%
+      mutate(Non.CO2 = "OC_AWB") -> OC_AWB_2000
+
+    # Aggregate BC and OC emissions to GCAM regions and convert to Tg
+    if(OLD_DATA_SYSTEM_BEHAVIOR) {
+      # Old system used match and ended up with just the first value instead of summing all values
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: 👍 nice catch!

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 4]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      bind_rows(OC_AWB_2000, BC_AWB_2000) %>%
+      left_join(iso_GCAM_regID, by = "iso") %>% #Not all iso's will necesarilly have an input value. This is ok.
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Can we use `left_join_error_no_match` here?

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 5]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      group_by(Non.CO2, GCAM_region_ID) %>%
+      summarise( awb = first( awb ) )  %>%
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: No GCAM-style spacing around parentheses and brackets please! (Throughout.)

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 6]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      filter( !is.na(awb) ) %>%
+      mutate( awb = awb * CONV_KG_TO_TG ) %>%
+      ungroup() ->
+      BCOC_AWB_2000
+    } else {
+        bind_rows(OC_AWB_2000, BC_AWB_2000) %>%
+        left_join(iso_GCAM_regID, by = "iso") %>% #Not all iso's will necesarilly have an input value. This is ok.
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Same question - `left_join_error_no_match` possible? Also, space after `#` please

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 7]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+        filter( !is.na(awb) ) %>%
+        mutate( awb = awb * CONV_KG_TO_TG ) %>%
+        ungroup() ->
+        BCOC_AWB_2000
+   }
+
+    # Extract only 2000 data from ag production data
+    filter( L103.ag_Prod_Mt_R_C_Y_GLU, year == 2000 ) %>%
+      select( -year ) %>%
+      mutate(GLU = as.character(GLU), GCAM_commodity = as.character(GCAM_commodity) ) ->
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Odd. They're not already character?

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 8]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      L103.ag_Prod_Mt_R_C_Y_GLU_2000
+
+    # Allocate aggreate regional year 2000 AWB emissions into GCAM crop and GLU
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: "aggregate"

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 9]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
+    # This is done using shares from L121.AWBshare_R_C_Y_GLU, which is calculated in another chunk
+    filter(L121.AWBshare_R_C_Y_GLU, year == "X2000")  %>%
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Ah...here's a significant change. This is currently `temp-data-inject/` data, i.e. temporarily spliced in from the old data system, but it _will_ be coming in from another chunk in long form. So won't have an "X2000" field name.

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 10]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
     # Load required inputs
     iso_GCAM_regID <- get_data(all_data, "common/iso_GCAM_regID")
     L103.ag_Prod_Mt_R_C_Y_GLU <- get_data(all_data, "L103.ag_Prod_Mt_R_C_Y_GLU")
-    L121.AWBshare_R_C_Y_GLU <- get_data(all_data, "L121.AWBshare_R_C_Y_GLU")
+    L121.AWBshare_R_C_Y_GLU <- get_data(all_data, "temp-data-inject/L121.AWBshare_R_C_Y_GLU")
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Immediately transform this data into eventual form (once not `temp-data-inject`): `get_data() %>% gather(year, value, -..., -...) %>% mutate(year = as.integer(substr(year, 2, 5)))`. Sorry if this wasn't clear. I may make this change.

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 11]
```diff
@@ -30,43 +30,78 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
     # Produce outputs
-    # Temporary code below sends back empty data frames marked "don't test"
-    # Note that all precursor names (in `add_precursor`) must be in this chunk's inputs
-    # There's also a `same_precursors_as(x)` you can use
-    # If no precursors (very rare) don't call `add_precursor` at all
-    tibble() %>%
-      add_title("descriptive title of data") %>%
+    L121.AWB_Emissions_FactorR_C_Y_GLU %>%
+      add_title("Emission factors for BC and OC emissions by GCAM GLU and crop") %>%
       add_units("units") %>%
```
Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Need to specify units

**Comment by bpbond** [comment]
So Steve - FYI I fixed this PR (the documentation) and then broke it (added a couple new lines transforming the `temp-data-inject` data).

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 4 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Not all iso have a value, so there is not always a match. The resulting NAs are then turned to zero, which is what we want if there wasn't data for that iso/species combination.

File: R/constants.R (hunk 1 above)
Comment (ssmithClimate) [inline R/constants.R]: Good idea, done.

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 6 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Same answer. :)

Comment (bpbond) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Thanks! Just checking. The more restrictive joins we use, the sooner any errors will be caught. -B

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 11 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Done.

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 9 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Got it.

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 2 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Done

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 12]
```diff
@@ -1,24 +1,24 @@
 #' module_emissions_L123.bcoc_awb_R_S_T_Y
 #'
-#' Briefly describe what this chunk does.
+#' Produces BC and OC AWB emission factors for each Ag production technology.
```
Comment (cahartin) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: what's AWB?

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 3 above)
Comment (cahartin) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: 👍

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R [hunk 13]
```diff
@@ -30,43 +30,80 @@ module_emissions_L123.bcoc_awb_R_S_T_Y_DISABLED <- function(command, ...) {
...
-    # If no precursors (very rare) don't call `add_precursor` at all
-    tibble() %>%
-      add_title("descriptive title of data") %>%
-      add_units("units") %>%
-      add_comments("comments describing how data generated") %>%
-      add_comments("can be multiple lines") %>%
+    L121.AWB_Emissions_FactorR_C_Y_GLU %>%
+      add_title("Emission factors for BC and OC emissions by GCAM GLU and crop") %>%
+      add_units("kt/Mt") %>%
+      add_comments("Derived from RCP year 2000 AWB emissions by region and allocated according to above ground non-harvested biomass") %>%
```
Comment (cahartin) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: can you add something like this comment up top?

File: R/zchunk_L123.bcoc_awb_R_S_T_Y.R (hunk 12 above)
Comment (ssmithClimate) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: Good point! Updated to define that. (Is agricultural waste burning on fields (AWB). This is specific burning of waste on fields, as distinguished from use of ag wastes for cooking or heating, which GCAM lumps into one type of biomass or another.)

Comment (cahartin) [inline R/zchunk_L123.bcoc_awb_R_S_T_Y.R]: cool!

**Pull request merged**